libazureinit/src/config.rs

```diff
@@ -0,0 +1,282 @@
+use toml;
+use tracing;
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy)]
+pub enum NetworkManager {
+    NetworkManager,
+    #[serde(rename = "systemd-networkd")]
+    #[default]
+    SystemdNetworkd,
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy)]
+pub enum SshAuthorizedKeysPathQueryMode {
+    #[serde(rename = "sshd -G")]
+    SshdG,
+    #[serde(rename = "disabled")]
+    #[default]
+    Disabled,
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+pub enum HostnameProvisioner {
+    #[default]
+    Hostnamectl,
+    #[cfg(test)]
+    FakeHostnamectl,
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+pub enum UserProvisioner {
+    #[default]
+    Useradd,
+    #[cfg(test)]
+    FakeUseradd,
+}
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+pub enum PasswordProvisioner {
+    #[default]
+    Passwd,
+    #[cfg(test)]
+    FakePasswd,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Network {
+    pub manage_configuration: bool,
+    pub network_manager: NetworkManager,
+}
+
+impl Default for Network {
+    fn default() -> Self {
+        Self {
+            manage_configuration: true,
+            network_manager: NetworkManager::SystemdNetworkd,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Ssh {
+    pub authorized_keys_path: PathBuf,
+    pub configure_password_authentication: bool,
+    pub authorized_keys_path_query_mode: SshAuthorizedKeysPathQueryMode,
+}
+
+impl Default for Ssh {
+    fn default() -> Self {
+        Self {
+            authorized_keys_path: PathBuf::from("~/.ssh/authorized_keys"),
+            configure_password_authentication: false,
+            authorized_keys_path_query_mode:
+                SshAuthorizedKeysPathQueryMode::Disabled,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct HostnameProvisioners {
+    pub backends: Vec<HostnameProvisioner>,
+}
+
+impl Default for HostnameProvisioners {
+    fn default() -> Self {
+        Self {
+            backends: vec![HostnameProvisioner::default()],
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct UserProvisioners {
+    pub backends: Vec<UserProvisioner>,
+}
+
+impl Default for UserProvisioners {
+    fn default() -> Self {
+        Self {
+            backends: vec![UserProvisioner::default()],
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct PasswordProvisioners {
+    pub backends: Vec<PasswordProvisioner>,
+}
+
+impl Default for PasswordProvisioners {
+    fn default() -> Self {
+        Self {
+            backends: vec![PasswordProvisioner::default()],
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Imds {
+    pub connection_timeout_secs: f64,
```

Comment:
why is connection timeout a float but read timeout unsigned int? They probably should all be float so that we can specify fraction of seconds if we want to